### PR TITLE
[layer] Support MoL attention layer forwarding

### PR DESCRIPTION
**Pull request by kparichay**
- This patch prepares scaffolding for MoL Attention Layer.
Attention layer has also been updated for MoLAttentionLayer to extend
AttentionLayer.
- This patch provides support for all the properties required for MoL
attention layer.
- This patch provides the forwarding implementation for the mol attention
layer. New underlying requirements have been added with this which
requires more tensor operations to support strided operations like
softmax which will be supported soon.
- add basic validation unittests + forwarding fix for mol
attention layer.
Further a validation test for mol layer is also added for larger size,
and nntrainer model is fixed to support multiple inputs.
- Add support for batched dot operaiton. Provide cleanup for both the
attention layers.\

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: nntrainer/layers/mol_attention_layer.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * @author Parichay Kapoor <pk.kapoor@samsung.com>
+ * @bug    No known bugs except for NYI items
+ * @brief  This is MoL Attention Layer Class for Neural Network
+ *
+ */
+
+#include <mol_attention_layer.h>
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+
```
Comment (zhoonit) [inline nntrainer/layers/mol_attention_layer.cpp:19]: please add <layer_context.h>, related to #1724

Comment (kparichay) [reply nntrainer/layers/mol_attention_layer.cpp:21]: done.

File: nntrainer/layers/mol_attention_layer.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  finalizeCommon(context);
+  /** NYI */
+}
+
+void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
+  /** NYI */
+}
+
+void MoLAttentionLayer::calcDerivative(RunLayerContext &context) { /** NYI */ }
+
```
Comment (jijoongmoon) [inline nntrainer/layers/mol_attention_layer.cpp:133]: Even if we are not going to implement the calcGradient for now, how about adding it?

**Review by jijoongmoon — APPROVED** [review]
LGTM!

File: nntrainer/layers/mol_attention_layer.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/layers/mol_attention_layer.cpp:133]: I will add.

File: nntrainer/layers/mol_attention_layer.cpp [hunk 3]
```diff
@@ -37,7 +38,8 @@ void MoLAttentionLayer::forwarding(RunLayerContext &context, bool training) {
 void MoLAttentionLayer::calcDerivative(RunLayerContext &context) { /** NYI */ }
 
 void MoLAttentionLayer::setProperty(const std::vector<std::string> &values) {
-  /** NYI */
+  auto remain_props = loadProperties(values, mol_props);
```
Comment (zhoonit) [inline nntrainer/layers/mol_attention_layer.cpp:135]: Could you also implement export_to as well?

File: nntrainer/layers/mol_attention_layer.cpp [hunk 4]
```diff
@@ -24,18 +25,103 @@ MoLAttentionLayer::~MoLAttentionLayer() {}
...
+  softmax.setActiFunc(ActivationType::ACT_SOFTMAX);
+  tanh.setActiFunc(ActivationType::ACT_TANH);
+  sigmoid.setActiFunc(ActivationType::ACT_SIGMOID);
+
+  auto unit = std::get<props::Unit>(mol_props).get();
+  auto mol_k = std::get<props::MoL_K>(mol_props).get();
+
+  TensorDim mlp_w_dim = {query_dim.width(), unit};
+  wt_idx[AttentionParams::mlp_w] =
+    context.requestWeight(mlp_w_dim, Tensor::Initializer::XAVIER_UNIFORM,
```
Comment (zhoonit) [inline nntrainer/layers/mol_attention_layer.cpp:58]: wonder if the initializers are fixed?

File: nntrainer/layers/mol_attention_layer.cpp [hunk 5]
```diff
@@ -24,18 +25,103 @@ MoLAttentionLayer::~MoLAttentionLayer() {}
...
+
+  /** @note kappa_hat, beta_hat, alpha_hat are strided */
+  Tensor kappa_hat = mlp_proj_out.getSharedDataTensor({batch, mol_k}, 0, false);
+  Tensor beta_hat =
+    mlp_proj_out.getSharedDataTensor({batch, mol_k}, mol_k, false);
+  Tensor alpha_hat =
+    mlp_proj_out.getSharedDataTensor({batch, mol_k}, mol_k * 2, false);
+
+  Tensor kappa = kappa_hat.apply(static_cast<float (*)(float)>(&std::exp));
+  Tensor beta = beta_hat.apply(static_cast<float (*)(float)>(&std::exp));
```
Comment (zhoonit) [inline nntrainer/layers/mol_attention_layer.cpp:99]: I guess we can use std::expf to avoid casting would happen inside function?

File: nntrainer/layers/acti_func.cpp [hunk 6]
```diff
@@ -156,6 +156,11 @@ Tensor &ActiFunc::softmax(Tensor const &t, Tensor &output) {
   float *dp;
   float *rp;
 
+  /** TODO: fix this check */
```
Comment (zhoonit) [inline nntrainer/layers/acti_func.cpp:159]: Could you elaborate on this comment? Is this enable strided softmax?

File: test/unittest/unittest_nntrainer_models.cpp [hunk 7]
```diff
@@ -741,6 +741,15 @@ INI multiout_model(
   }
 );
 
+INI mol_attention(
+  "mol_attention",
+  {nn_base + "batch_size = 3",
+   sgd_base + "learning_rate = 1",
+   I("in0") + input_base + "input_shape = 1:1:256",
+   I("in1") + input_base + "input_shape = 1:84:256",
+   I("in2") + input_base + "input_shape = 1:1:5", // this shape should match mol_k
```
Comment (zhoonit) [inline test/unittest/unittest_nntrainer_models.cpp:750]: If shape should match mol_k, can we use input_dim.width() to infer mol_k instead of setting it?

File: nntrainer/layers/mol_attention_layer.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/layers/mol_attention_layer.cpp:135]: it coming in the new PR.

File: nntrainer/layers/mol_attention_layer.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/layers/mol_attention_layer.cpp:58]: Its updated to use LayerImpl in the upcoming PR.

File: nntrainer/layers/mol_attention_layer.cpp (hunk 5 above)
Comment (kparichay) [reply nntrainer/layers/mol_attention_layer.cpp:99]: Will use expf (not std::expf) in upcoming PR.

File: test/unittest/unittest_nntrainer_models.cpp (hunk 7 above)
Comment (kparichay) [reply test/unittest/unittest_nntrainer_models.cpp:750]: It seems like mol_k should be able to interpret, but has been kept as a parameter as per the reference python implementation. If it is always fixed, will remove the property and interpret it from the input dimension later.

File: nntrainer/layers/acti_func.cpp (hunk 6 above)
Comment (kparichay) [reply nntrainer/layers/acti_func.cpp:159]: Yes, it is to enable strided softmax in future. Updated the comment.

File: test/unittest/unittest_nntrainer_models.cpp (hunk 7 above)
Comment (zhoonit) [reply test/unittest/unittest_nntrainer_models.cpp:750]: I see, thanks for the explanation :)

**Review by zhoonit — APPROVED** [review]
LGTM
